### ISSUE TITLE
[BUG] ZeRO-Offload GPU -> CPU datatype

### ISSUE THREAD

**Issue by taehyunzzz** [microsoft/DeepSpeed#2986]
From my understanding of ZeRO-Offload mixed-precision training, it is communication-efficient because datatype of gradient data and updated weights transferred between CPU and GPU is FP16 instead of FP32. However, [this part](https://github.com/microsoft/DeepSpeed/blob/58a4a4d4c19bda86d489ac171fa10f3ddb27c9d6/deepspeed/runtime/zero/stage_1_and_2.py#L1187) suggests that FP16 gradient is converted to FP32 in GPU, which is then transferred to CPU. I think that there should be a FP16 buffer on the CPU side which is used to buffer the parameter gradients sent from the GPU, whose data should then be converted to FP32 destination buffer. I thought this could be a bug, but is there a reason for this? 

The below test code shows "CUDA FP16 -> CPU FP16 -> CPU FP32" is faster than "CUDA FP16 -> CUDA FP32 -> CPU FP32". Copying a FP16 CUDA tensor directly into FP32 CPU tensor almost showed the same speed as "CUDA FP16 -> CPU FP16 -> CPU FP32".

Output was : 

![image](https://user-images.githubusercontent.com/33646149/224122811-0ca67787-4be8-4c2d-8635-c3dd74c9ba1a.png)

```
import torch
import time

cpu_fp16 = torch.rand([500000], pin_memory=True, dtype=torch.float16)
cpu_fp32 = torch.rand([500000], pin_memory=True, dtype=torch.float32)
gpu_fp16 = torch.rand([500000], pin_memory=True, dtype=torch.float16)
gpu_fp32 = torch.rand([500000], pin_memory=True, dtype=torch.float32)

testcase = 10000

# First case : CUDA FP16 -> CUDA FP32 -> CPU FP32
t1= time.time()
for i in range(testcase):
    gpu_fp16 = gpu_fp16.float()
    cpu_fp32.copy_(gpu_fp16)
t2 = time.time()

cpu_fp16 = torch.rand([500000], pin_memory=True, dtype=torch.float16)
cpu_fp32 = torch.rand([500000], pin_memory=True, dtype=torch.float32)
gpu_fp16 = torch.rand([500000], pin_memory=True, dtype=torch.float16)
gpu_fp32 = torch.rand([500000], pin_memory=True, dtype=torch.float32)

# Second case : CUDA FP16 -> CPU FP16 -> CPU FP32
t3= time.time()
for i in range(testcase):
    cpu_fp16.copy_(gpu_fp16)
    cpu_fp16 = cpu_fp16.float()
t4 = time.time()

cpu_fp16 = torch.rand([500000], pin_memory=True, dtype=torch.float16)
cpu_fp32 = torch.rand([500000], pin_memory=True, dtype=torch.float32)
gpu_fp16 = torch.rand([500000], pin_memory=True, dtype=torch.float16)
gpu_fp32 = torch.rand([500000], pin_memory=True, dtype=torch.float32)

# Third case : CUDA FP16 -> CPU FP32
t5= time.time()
for i in range(testcase):
    cpu_fp32.copy_(gpu_fp16)
t6 = time.time()

print("EXP1 {}\nEXP2 {}\nEXP3 {}".format(t2-t1, t4-t3, t6-t5))

```

**Comment by tjruwase** [comment]
@taehyunzzz, thanks for sharing your analysis. If I understand correctly, it seems that `CUDA FP16 -> CPU FP32` would be an improvement. Do you know if this this feature is supported by older pytorch versions? Also, do you mind providing a PR for this?

**Comment by taehyunzzz** [comment]
@tjruwase Opened [PR#3013](https://github.com/microsoft/DeepSpeed/pull/3013#issue-1623263589). Please review :)

**Comment by taehyunzzz** [comment]
@tjruwase Oh, I did not answer the first question. No, I do not know if older pytorch versions support this at the moment. I think this function has to do with aten library as well, so this should be considered as well.

**Comment by taehyunzzz** [comment]
Actually, I found that when I run the training code the current version is faster. 

Trace shows that for the original code, src_tensor.float() -> dst_tensor.copy_(src_tensor, non_blocking=True) is executed asynchronously (actual GPU kernel is ran seconds later). In the image below, "aten:to" corresponds to src_tensor.float() and aten::copy_ corresponds to dest_tensor.copy_(src_tensor)
![image](https://user-images.githubusercontent.com/33646149/224995037-cf584659-bb32-40f1-bdd0-a0ff284332d0.png)

The change I made (removing the src_tensor.float() code) undergoes so much slowdown because data is sent asynchronously but the conversion from FP16 to FP32 in CPU seems to block the whole execution. Here, copy_ is executed in 2 steps, wherein data movement from GPU to CPU is done in the first aten::copy_ while type conversion is done in the second aten::copy_.
![image](https://user-images.githubusercontent.com/33646149/224996656-81ddc1b0-3c56-4e8c-922c-2d1913d0ea89.png)

Using FP16 CPU buffer to copy the values from GPU to CPU asynchronously, and deferring the type conversion should alleviate the communication overhead, but the memory space and extra computation may worsen performance... :(

Conclusion : I think the code works better as is for now, but there is room for improvement.
